pkg/controllers/selection/controller.go

```diff
@@ -84,18 +84,33 @@ func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reco
 }
 
 func (c *Controller) selectProvisioner(ctx context.Context, pod *v1.Pod) (errs error) {
-	// Relax preferences if pod has previously failed to schedule.
-	c.preferences.Relax(ctx, pod)
-	// Inject volume topological requirements
-	if err := c.volumeTopology.Inject(ctx, pod); err != nil {
-		return fmt.Errorf("getting volume topology requirements, %w", err)
-	}
 	// Pick provisioner
 	var provisioner *provisioning.Provisioner
 	provisioners := c.provisioners.List(ctx)
 	if len(provisioners) == 0 {
+		// checking for existing provisioners first prevents relaxing pods which logs messages that make it appear
+		// Karpenter is attempting to schedule even when you don't have a provisioner created.  This also prevents an issue
+		// where if you created a provisioner later, the pod may have already been fully relaxed unnecessarily.
 		return nil
 	}
+
+	// We only add the toleration for a PreferNoSchedule taint if one of the provisioners puts this taint on. Otherwise
+	// it would have no effect and produces a confusing log message.
+	hasPreferNoScheduleTaint := false
+	for _, prov := range provisioners {
+		for _, taint := range prov.Spec.Taints {
+			if taint.Effect == v1.TaintEffectPreferNoSchedule {
+				hasPreferNoScheduleTaint = true
+			}
+		}
+	}
+
+	// Relax preferences if pod has previously failed to schedule.
+	c.preferences.Relax(ctx, pod, hasPreferNoScheduleTaint)
```

Comment:
If we would expand these conditions later, it might be cleaner to pass the provisioner to the Relax func and throw the `hasPreferNoScheduleTaint` into a func. Wdyt?

Reply:
I thought about that, but it's actually a list of provisioners that we would need to pass in.  That seemed a bit more weird to me, but I'm good either way.  Back to you, Wdyt? 😂

Reply:
I think the slice of provisioners is fine to pass to `Relax()`. To me, it's cleaner to think about Relax taking into account the things passed to it rather than exposing that directly in the `selectProvisioner()` and passing a flag.

Reply:
Turned it into a generic options struct, seems to make more sense that way.  We can expand it if needed without making preferences aware of provisioners.